Add docstring for saved tensors default hooks

Varal7 opened the pull request:
Add documentation for the saved tensors default hooks introduced in #61834 / #62563

Sister PR: #62362 (will add a link from autograd.rst to notes/autograd in whatever PR does not land first)

Varal7 commented on torch/autograd/saved_variable_default_hooks.py:
```diff
@@ -1,13 +1,85 @@
+    r"""Sets a pair of pack / unpack hooks for saved tensors.
+
+    Use these hooks to define how intermediary results of an operation
+    should be packed before saving, and unpacked on retrieval.
```
Here is where I would add a link to https://github.com/pytorch/pytorch/pull/62362 which defines the terms `packing` and `unpacking`.

Varal7 commented on torch/autograd/__init__.py:
```diff
@@ -20,7 +20,7 @@
 from ..overrides import has_torch_function, handle_torch_function
 from . import functional
 from . import forward_ad
-from . import saved_variable_default_hooks as graph
```
Somehow autodoc can't find the `graph` module if it's imported
with an alias.
https://www.internalfb.com/intern/opensource/ci/job/log/265181241677988/

zou3519 commented on docs/source/autograd.rst:
```diff
@@ -241,3 +241,19 @@ Anomaly detection
+    :toctree: generated
+    :nosignatures:
+
+    graph.set_saved_tensors_default_hooks
+    graph.reset_saved_tensors_default_hooks
```
Hmm, I'd argue that these should be context managers especially if we are going to change the default hooks to be thread-local. I'm fine with having a function-only option in the future, but only if we heavily advertise the context manager version (for safety) like we do with torch.set_grad_enabled

Varal7 committed:
Add docstring for saved tensors default hooks

Adds documentation for the saved tensors default hooks introduced in #61834

ghstack-source-id: aa1104694c41ca2e609480e6623b3d03de44fa37
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62922

Varal7 committed:
Rename saved_variable_default_hooks into graph

Somehow autodoc can't find the `graph` module if it's imported
with an alias.
https://www.internalfb.com/intern/opensource/ci/job/log/265181241677988/

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ghstack-source-id: 2a115fdb9e1d6d5537b7015b19739ee11efa497b
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62923

Varal7 committed:
Change API of default hooks to be set with a context-manager

ghstack-source-id: d316e4f5419428eabe4086aad1be6b597a2c9f24
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62924

zou3519 commented on docs/source/autograd.rst:
```diff
@@ -241,3 +241,15 @@ Anomaly detection
+
+Some operations need intermediary results to be saved during the forward pass
+in order to execute the backward pass.
+You can define how these saved tensors should be packed / unpacked using hooks.
```
Some more motivation here would be nice. E.g. "you might want to reduce memory if your model is just slightly bigger than the GPU memory you're using", "you might want to swap out to disk", etc.

zou3519 commented on torch/autograd/graph.py:
```diff
@@ -0,0 +1,139 @@
+
+    where the return value of ``pack_hook`` is a valid input to ``unpack_hook``.
+
+    In general, you want ``unpack_hook(pack_hook(t))`` to be equal to ``t`` in terms
+    of value, size and dtype.
```
what about device?

zou3519 commented on torch/autograd/graph.py:
```diff
@@ -0,0 +1,139 @@
+        torch._C._autograd._reset_default_hooks()
+
+
+class save_on_cpu():
```
nit: you might be able to implement `save_on_cpu` in terms of `saved_tensors_hooks` but it doesn't save too many LOC
